packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -54,12 +54,16 @@ function getChildResultsRecursively(
 function resolveChildrenFromFields(
   parent: ResultWithFolding,
   results: ResultWithFolding[],
-  fields: FoldingFields
+  fields: FoldingFields,
+  resolvedAncestors: string[] = []
 ): FoldedResult[] {
   const sourceChildValue = getChildField(parent, fields);
   if (!sourceChildValue) {
     return [];
   }
+  if (resolvedAncestors.indexOf(sourceChildValue) !== -1) {
+    return [];
+  }
```

Comment:
I'm not sure how the API would behave if there's a circular reference, but either way I think we should add some check against a recursive loop.